zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -229,83 +259,99 @@ impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
-                let rpi_field_bytes_acc_next =
-                    meta.query_advice(rpi_field_bytes_acc, Rotation::next());
-                let rpi_field_bytes_acc = meta.query_advice(rpi_field_bytes_acc, Rotation::cur());
-                let rpi_field_bytes_next = meta.query_advice(rpi_field_bytes, Rotation::next());
-                let is_field_rlc = meta.query_fixed(is_field_rlc, Rotation::next());
-                let randomness = challenges.evm_word();
-                let t = select::expr(is_field_rlc, randomness, BYTE_POW_BASE.expr());
-                cb.require_equal(
-                    "rpi_field_bytes_acc[i+1] = rpi_field_bytes_acc[i] * t + rpi_bytes[i+1]",
-                    rpi_field_bytes_acc_next,
-                    rpi_field_bytes_acc * t + rpi_field_bytes_next,
-                );
-                cb.gate(q_field_step)
-            },
-        );
-        meta.create_gate("rpi_field_bytes_acc[0] = rpi_field_bytes[0]", |meta| {
-            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
-
-            let q_field_start = meta.query_selector(q_field_start);
-            let rpi_field_bytes_acc = meta.query_advice(rpi_field_bytes_acc, Rotation::cur());
-            let rpi_field_bytes = meta.query_advice(rpi_field_bytes, Rotation::cur());
-
-            cb.require_equal(
-                "rpi_field_bytes_acc[0] = rpi_field_bytes[0]",
-                rpi_field_bytes_acc,
-                rpi_field_bytes,
-            );
-            cb.gate(q_field_start)
+            circuit!([meta, cb], {
+                // rpi_field_bytes_acc[i+1] = rpi_field_bytes_acc[i] * t + rpi_bytes[i+1]
+                ifx!(q!(q_field_step) => {
+                    let t = ifx!(f!(is_field_rlc) => {
+                        challenges.evm_word()
+                    } elsex {
+                        BYTE_POW_BASE.expr()
+                    });
+                    require!(
+                        a!(rpi_field_bytes_acc, 1) => a!(rpi_field_bytes_acc) * t + a!(rpi_field_bytes, 1)
+                    );
+                });
+                // rpi_field_bytes_acc[0] = rpi_field_bytes[0]
+                ifx!(q!(q_field_start) => {
+                    require!(a!(rpi_field_bytes_acc) => a!(rpi_field_bytes));
+                });
+            });
+            cb.build_constraints(None)
         });
 
-        // keccak in rpi
-        meta.lookup_any("keccak(rpi)", |meta| {
-            let q_keccak = meta.query_selector(q_keccak);
-            let rpi_rlc = meta.query_advice(rpi_field_bytes_acc, Rotation::cur());
-            let output = meta.query_advice(rpi_rlc_acc, Rotation::cur());
-            [1.expr(), rpi_rlc, RPI_BYTES_LEN.expr(), output]
-                .into_iter()
-                .zip(keccak_table.table_exprs(meta).into_iter())
-                .map(|(arg, table)| (q_keccak.expr() * arg, table))
-                .collect::<Vec<_>>()
+        meta.create_gate("keccak(rpi)", |meta| {
+            cb.restart();
```

Comment:
Much cleaner to just move everything in a single gate like the MPT circuit, then also no restart necessary (and I guess not always restarted now, so constraints are added multiple times now per gate because they accumulate)?

Reply:
Glad that you like it :) No restart needed.
What do you mean accumulate tho? If you're talking about `block_acc` it should be applied per row within the left field bytes section.

Reply:
I meant that the multiple gates introduces extra things people can forget and make errors with. Because each gate does `build_constraints`, if you don't call `restart` between the different gates, constraints will keep getting added to the list in the constrain builder and those will keep being added when calling `build_constraints`, so that will also add the constraints in the previous gates again.

Conclusion: gates bad! haha